RecoTracker/TkSeedGenerator/plugins/SeedingOTEDProducer.cc

```diff
@@ -0,0 +1,392 @@
+  edm::ESGetToken<TrajectoryStateUpdator, TrackingComponentsRecord> updatorToken_;
+  edm::ESGetToken<MeasurementTracker, CkfComponentsRecord> measurementTrackerToken_;
+  edm::ESGetToken<Chi2MeasurementEstimatorBase, TrackingComponentsRecord> estToken_;
+};
+
+SeedingOTEDProducer::SeedingOTEDProducer(edm::ParameterSet const& conf)
+    : updator_(nullptr),
+      tkMeasEventToken_(consumes<MeasurementTrackerEvent>(conf.getParameter<edm::InputTag>("trackerEvent"))),
+      topoToken_(esConsumes()),
+      propagatorToken_(esConsumes(edm::ESInputTag("", "PropagatorWithMaterial"))),
+      magFieldToken_(esConsumes()),
+      updatorToken_(esConsumes()),
+      measurementTrackerToken_(esConsumes()),
+      estToken_(esConsumes(edm::ESInputTag("", "Chi2"))) {
+  vhProducerToken_ = consumes<VectorHitCollectionNew>(edm::InputTag(conf.getParameter<edm::InputTag>("src")));
+  beamSpotToken_ = consumes<reco::BeamSpot>(conf.getParameter<edm::InputTag>("beamSpotLabel"));
+  updatorName_ = conf.getParameter<std::string>("updator");
+  produces<TrajectorySeedCollection>();
+}
+
+SeedingOTEDProducer::~SeedingOTEDProducer() {}
+
+void SeedingOTEDProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+  edm::ParameterSetDescription desc;
+  desc.add<edm::InputTag>("src", edm::InputTag("siPhase2VectorHits", "vectorHitsAccepted"));
+  desc.add<edm::InputTag>("trackerEvent", edm::InputTag("MeasurementTrackerEvent"));
+  desc.add<edm::InputTag>("beamSpotLabel", edm::InputTag("offlineBeamSpot"));
+  desc.add<std::string>("updator", std::string("KFUpdator"));
+  descriptions.add("SeedingOTEDProducer", desc);
+}
+
+void SeedingOTEDProducer::produce(edm::Event& event, const edm::EventSetup& es) {
+  std::unique_ptr<TrajectorySeedCollection> seedsWithVHs(new TrajectorySeedCollection());
+
+  tkTopo_ = &es.getData(topoToken_);
+
+  edm::ESHandle<MeasurementTracker> measurementTrackerHandle;
+  measurementTrackerHandle = es.getHandle(measurementTrackerToken_);
+  measurementTracker_ = measurementTrackerHandle.product();
+
+  edm::Handle<MeasurementTrackerEvent> measurementTrackerEvent;
+  event.getByToken(tkMeasEventToken_, measurementTrackerEvent);
+
+  LayerMeasurements layerMeasurements_(*measurementTrackerHandle, *measurementTrackerEvent);
+
+  estimator_ = &es.getData(estToken_);
+
+  propagator_ = &es.getData(propagatorToken_);
+
+  magField_ = &es.getData(magFieldToken_);
+
+  updator_ = &es.getData(updatorToken_);
+
+  edm::Handle<reco::BeamSpot> beamSpotH;
+  event.getByToken(beamSpotToken_, beamSpotH);
+  if (beamSpotH.isValid()) {
+    beamSpot_ = beamSpotH.product();
+  }
+
+  // Get the vector hits
+  edm::Handle<VectorHitCollectionNew> vhs;
+  event.getByToken(vhProducerToken_, vhs);
+
+  TrajectorySeedCollection const& tempSeeds = run(vhs);
+  for (TrajectorySeedCollection::const_iterator qIt = tempSeeds.begin(); qIt < tempSeeds.end(); ++qIt) {
+    seedsWithVHs->push_back(*qIt);
+  }
```

Comment:
Couldn't this loop be removed altogether?
```cpp
auto seedsWithVHs = std::make_unique<TrajectorySeedCollection>(tempSeeds);
```
should do the same.

In fact, with an `edm::EDPutTokenT` (`edm::EDPutTokenT<TrajectorySeedCollection>` member, initialized with the return value of the `produces<TrajectorySeedCollection>()` call), the code could be simplified to
```cpp
auto tempSeeds = run(vhs);
tempSeeds.shrink_to_fit();
event.put(putToken_, std::move(tempSeeds));
```
or even
```cpp
// shrink_to_fit() needs to be called in run()
event.put(putToken_, run(vhs));
```

Reply:
Needed a bit of rearranging to have `run` return a `unique_ptr`, but I have implemented the last option you suggested.

Reply:
Apologies, I'm confusing myself. What I really meant with both `edm::EDPutTokenT` recipies was
```cpp
event.emplace(putToken_, run(vhs));
```
in which case `run()` could return just `TrajectorySeedCollection` (avoiding one `std::unique_ptr`). I leave it to your discretion if this change is worth it or not.

Reply:
I have adopted this change.